src/generator_files.rs

```diff
@@ -129,45 +114,42 @@ fn ping_files(
 }
 
 pub fn orchestrate(client: &DatagoClient) -> DatagoEngine {
-    // Start pulling the samples, which spread across three steps. The samples will land in the last kanal,
+    // Start pulling the samples, which spread across two steps. The samples will land in the last kanal,
     // all the threads pausing when the required buffer depth is reached.
 
     // A first thread will query the filesystem and get pages of filepaths back
     // This meta data is then dispatched to a worker pool, which will load the files, deserialize them,
     // do the required pre-processing then commit to the ready queue.
 
-    // TODO: refactor to join with the same orchestration function in generator_http.rs
+    // TODO: Pass over an Arc ref of the client instead of doing the current member copies
 
     // Allocate all the message passing pipes
-    let (pages_tx, pages_rx) = bounded(2);
     let (samples_metadata_tx, samples_metadata_rx) = bounded(client.samples_buffer * 2);
     let (samples_tx, samples_rx) = bounded(client.samples_buffer);
 
     // Convert the source_config to a SourceFileConfig
     let source_config: SourceFileConfig =
         serde_json::from_value(client.source_config.clone()).unwrap();
-
     println!("Using file as source {}", source_config.root_path);
+
+    // Create a thread which will generate work as it goes. We'll query the filesystem
+    // and send the filepaths to the worker pool as we go
     let rank = client.rank;
     let limit = client.limit;
     let world_size = client.world_size;
 
-    let pinger = Some(thread::spawn(move || {
```

Comment:
gist of this PR, one less thread, list files and dispatch them to the workers directly